pkg/operator/osd/osd.go

```diff
@@ -16,69 +16,136 @@ limitations under the License.
 package osd
 
 import (
+	"bytes"
 	"fmt"
 
+	"strings"
+
 	"github.com/rook/rook/pkg/operator/k8sutil"
 	opmon "github.com/rook/rook/pkg/operator/mon"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/pkg/api/errors"
+	"k8s.io/client-go/pkg/api/unversioned"
 	"k8s.io/client-go/pkg/api/v1"
 	extensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
 )
 
 const (
-	appName = "osd"
+	appName    = "osd"
+	appNameFmt = "osd-%s"
 )
 
 type Cluster struct {
 	clientset       kubernetes.Interface
 	Namespace       string
 	Keyring         string
 	Version         string
-	useAllDevices   bool
+	Storage         StorageSpec
 	dataDirHostPath string
-	deviceFilter    string
 }
 
-func New(clientset kubernetes.Interface, namespace, version, deviceFilter, dataDirHostPath string, useAllDevices bool) *Cluster {
+func New(clientset kubernetes.Interface, namespace, version string, storageSpec StorageSpec, dataDirHostPath string) *Cluster {
 	return &Cluster{
 		clientset:       clientset,
 		Namespace:       namespace,
 		Version:         version,
-		deviceFilter:    deviceFilter,
+		Storage:         storageSpec,
 		dataDirHostPath: dataDirHostPath,
-		useAllDevices:   useAllDevices,
 	}
 }
 
 func (c *Cluster) Start() error {
-	logger.Infof("start running osds")
-
-	ds, err := c.makeDaemonSet()
-	_, err = c.clientset.Extensions().DaemonSets(c.Namespace).Create(ds)
-	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create osd daemon set. %+v", err)
+	logger.Infof("start running osds in namespace %s", c.Namespace)
+
+	if c.Storage.UseAllNodes {
+		// make a daemonset for all nodes in the cluster
+		ds := c.makeDaemonSet(c.Storage.Selection, c.Storage.Config)
+		_, err := c.clientset.Extensions().DaemonSets(c.Namespace).Create(ds)
+		if err != nil {
+			if !errors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create osd daemon set. %+v", err)
+			}
+			logger.Infof("osd daemon set already exists")
+		} else {
+			logger.Infof("osd daemon set started")
 		}
-		logger.Infof("osd daemon set already exists")
 	} else {
-		logger.Infof("osd daemon set started")
+		for i := range c.Storage.Nodes {
```

Comment:
That's correct, a lot of these params will go away in a later iteration when the API server will hand out the config to each OSD when they ask for it.